wallet/store.go

```diff
@@ -1,157 +1,242 @@
+		if k.salt != nil {
+			log.Fatalf("Can only set salt once.")
+		}
+		k.salt = make([]byte, Pbkdf2SaltBytesLen)
 		_, err := rand.Read(k.salt)
 		cobra.CheckErr(err)
-		k.key = pbkdf2.Key([]byte(password), k.salt,
-			Pbkdf2Itterations,
-			EncKeyLen,
-			Pbkdf2HashFunc,
-		)
 	}
 }
 
-// Is better, but not FIPS-140 compliant.
-func WithArgon2idPassword(password string) WalletKeyOpt {
+func WithSalt(salt [Pbkdf2SaltBytesLen]byte) WalletKeyOpt {
 	return func(k *WalletKey) {
-		hash, err := argon2id.CreateHash(password, Argon2idParams)
-		cobra.CheckErr(err)
-		_, salt, key, err := argon2id.DecodeHash(hash)
-		cobra.CheckErr(err)
-		k.salt = salt
-		k.key = key
+		if k.salt != nil {
+			log.Fatalf("Can only set salt once.")
+		}
+		k.salt = salt[:]
+
+		// if password is set, set the key as well
+		if k.pw != nil {
+			WithPbkdf2Password(k.pw)(k)
+		}
+	}
+}
+
+func WithIterations(iterations int) WalletKeyOpt {
+	return func(k *WalletKey) {
+		k.iterations = iterations
+		if k.key != nil {
+			// regenerate
+			k.key = nil
+			WithPbkdf2Password(k.pw)(k)
+		}
+	}
+}
+
+// WithPasswordOnly is used for reading a stored file. The stored wallet file contains
+// a salt, so it does not need to be set before reading the file.
+func WithPasswordOnly(password []byte) WalletKeyOpt {
+	return func(k *WalletKey) {
+		if k.salt != nil {
+			log.Fatalf("Salt must not be set.")
+		}
+		if k.key != nil {
+			log.Fatalf("Can only generate key once.")
+		}
+		if k.pw != nil {
+			log.Fatalf("Password can only be set once.")
+		}
+		k.pw = password
+	}
+}
+
+func WithPbkdf2Password(password []byte) WalletKeyOpt {
+	return func(k *WalletKey) {
+		if k.salt == nil {
+			log.Fatalf("Salt must be set.")
+		}
+		if k.key != nil {
+			log.Fatalf("Can only generate key once.")
+		}
+		iterations := k.iterations
+		if iterations == 0 {
+			iterations = Pbkdf2Iterations
+		}
+		k.key = pbkdf2.Key(
+			password,
+			k.salt,
+			iterations,
+			EncKeyLen,
+			Pbkdf2HashFunc,
+		)
+		k.pw = password
 	}
 }
 
 // https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#71-encryption-types-to-use
 func (k *WalletKey) encrypt(plaintext []byte) (ciphertext []byte, nonce []byte) {
 	block, err := aes.NewCipher(k.key)
 	cobra.CheckErr(err)
-	// Never use more than 2^32 random nonces with a given key because of the risk of a repeat.
-	// TODO: Remind user to not use the same password to encrypt 4 billion times.
-	nonce = make([]byte, 12)
-	_, err = io.ReadFull(rand.Reader, nonce)
-	cobra.CheckErr(err)
+
+	// Using default options for AES-GCM as recommended by the godoc.
+	// For reference, NonceSize is 12 bytes, and TagSize is 16 bytes:
+	// https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/crypto/cipher/gcm.go;l=153-158
 	aesgcm, err := cipher.NewGCM(block)
+	nonce = make([]byte, aesgcm.NonceSize())
+	hash := hmac.New(sha512.New, k.key)
+	hash.Write(plaintext)
+	nonce = hash.Sum(nil)[:aesgcm.NonceSize()]
```

Comment:
nonce is initialized and immediately replaced, the re-slicing of the hash should also not be required:

```suggestion
	hash := hmac.New(sha512.New, k.key)
	nonce := hash.Sum(plaintext)
```

Reply:
I also find it a bit odd, that the nonce - which should be a random value - is derived from the plaintext.

Reply:
The only requirement in AES-GCM is that the same nonce not be used twice for two different messages. It does not have to be random.